contrib/verifybinaries/verify.py

```diff
@@ -47,136 +189,498 @@ def parse_version_string(version_str):
     return version_base, version_rc, version_os
 
 
-def download_with_wget(remote_file, local_file=None):
-    if local_file:
-        wget_args = ['wget', '-O', local_file, remote_file]
-    else:
-        # use timestamping mechanism if local filename is not explicitly set
-        wget_args = ['wget', '-N', remote_file]
-
-    result = subprocess.run(wget_args,
+def download_with_wget(remote_file, local_file):
+    result = subprocess.run(['wget', '-O', local_file, remote_file],
                             stderr=subprocess.STDOUT, stdout=subprocess.PIPE)
     return result.returncode == 0, result.stdout.decode().rstrip()
 
 
+def download_lines_with_urllib(url) -> (bool, [str]):
+    """Get (success, text lines of a file) over HTTP."""
+    try:
+        return True, [
+            line.strip().decode() for line in urllib.request.urlopen(url).readlines()]
+    except urllib.request.HTTPError as e:
+        log.warning(f"HTTP request to {url} failed (HTTPError): {e}")
+    except Exception as e:
+        log.warning(f"HTTP request to {url} failed ({e})")
+    return (False, [])
+
+
+def verify_with_gpg(
+    signature_filename,
+    output_filename: t.Optional[str] = None
+) -> t.Tuple[int, str]:
+    args = [
+        'gpg', '--yes', '--decrypt',
+        '--output', output_filename if output_filename else '', signature_filename]
+
+    env = dict(os.environ, LANGUAGE='en')
+    result = subprocess.run(args, stderr=subprocess.STDOUT, stdout=subprocess.PIPE, env=env)
+    log.debug(f'Result from GPG ({result.returncode}): {result.stdout}')
+    return result.returncode, result.stdout.decode().rstrip()
+
+
+def remove_files(filenames):
+    for filename in filenames:
+        os.remove(filename)
+
+
+class SigData:
+    """GPG signature data as parsed from GPG stdout."""
+    def __init__(self, key: str, name: str, trusted: bool, status: str):
+        self.key = key
+        self.name = name
+        self.trusted = trusted
+        self.status = status
+
+    def __repr__(self):
+        return (
+            "SigData(%r, %r, trusted=%s, status=%r)" %
+            (self.key, self.name, self.trusted, self.status))
+
+
+def parse_gpg_result(output: str) -> ([SigData], [SigData], [SigData]):
+    """Returns good, unknown, and bad signatures from GPG stdout."""
+    good_sigs = []
+    unknown_sigs = []
+    bad_sigs = []
+    total_resolved_sigs = 0
+    curr_key = None
+
+    # Ensure that all lines we match on include a prefix that prevents malicious input
+    # from fooling the parser.
+    def line_begins_with(patt: str, line: str) -> bool:
+        return re.match(r'^\s*(gpg:)?(\s+)' + patt, line)
+
+    for i, line in enumerate(output):
+        if line_begins_with(r"using RSA key (0x[0-9a-fA-F]{16}|[0-9a-fA-F]{40})$", line):
```

Comment:
For v23.0 one user now has an ECDSA key which seems to break this script (`output[70:80]`)
  ```
  gpg: Signature made Sat 23 Apr 2022 13:39:56 BST
  gpg:                using ECDSA key C388F6961FB972A95678E327F62711DBDCA8AE56
  gpg:                issuer "kvaciral@protonmail.com"
  gpg: key 0x2F9867672542AC32: "kvaciral@protonmail.com <kvaciral@protonmail.com>" not changed
  gpg: Total number processed: 1
  gpg:              unchanged: 1
  gpg: requesting key 0xF62711DBDCA8AE56 from hkps server keys.openpgp.org
  gpg: key 0xF62711DBDCA8AE56: new key but contains no user ID - skipped
  gpg: Total number processed: 1
  gpg:           w/o user IDs: 1
  gpg: Can't check signature: No public key
  ```
  
  After this block is skipped `curr_key` is not set, which then causes the `RuntimeError` on L269 to hit